allennlp/modules/seq2seq_decoders/seq_decoder.py

```diff
@@ -0,0 +1,38 @@
+from typing import Dict
+
+import torch
+from torch.nn import Module
+
+from allennlp.common import Registrable
+from allennlp.modules import Embedding
+
+
+class SeqDecoder(Module, Registrable):
+    """
+    A ``SeqDecoder`` is a base class for different types of Seq decoding modules
+    Parameters
+    ----------
+    target_embedder : ``Embedding``
+        Embedder for target tokens. Needed in the base class to enable weight tying.
+    """
+    default_implementation = 'auto_regressive_seq_decoder'
+
+    def __init__(self,
+                 target_embedder: Embedding) -> None:
+        super().__init__()
+        self.target_embedder = target_embedder
+
+    def get_output_dim(self):
+        raise NotImplementedError()
+
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        raise NotImplementedError()
+
+    def forward(self,
```

Comment:
Since this is an abstract class it's very important that functions have comments describing their purpose.